Makefile

```diff
@@ -222,29 +224,53 @@ docker-hotrod:
 	docker build -t $(DOCKER_NAMESPACE)/example-hotrod:${DOCKER_TAG} ./examples/hotrod --build-arg TARGETARCH=$(GOARCH)
 
 .PHONY: run-all-in-one
-run-all-in-one:
+run-all-in-one: build-ui
 	go run -tags ui ./cmd/all-in-one --log-level debug
 
+# Be explicit about exactly what assets we want, as asset files could be partially deleted
+# after initial generation; which is the case in Travis CI builds (DOCKER + DEPLOY).
+EXPECT_ASSETS := cmd/query/app/ui/actual/gen_assets.go \
+                 cmd/query/app/ui/placeholder/gen_assets.go
+FOUND_ASSETS := $(wildcard $(EXPECT_ASSETS))
 .PHONY: build-ui
 build-ui:
+	# The `jaeger-ui` submodule contains the source code for the UI assets (requires Node.js 6+).
+	# The assets must be compiled first with `make build-ui`, which runs Node.js build and then
+	# packages the assets into a Go file that is `.gitignore`-ed.
+	#
+	# The packaged assets can be enabled by providing a build tag `ui`; for example:
+	# $ go run -tags ui ./cmd/all-in-one/main.go
+	#
+	# To reduce `build-ui` execution times, no work will be done if the resulting UI assets exist.
+	# These expected assets are cmd/query/app/ui/(actual|placeholder)/gen_assets.go.
+	#
+	# To force a rebuild of UI assets, run `make clean` which deletes any files matching the
+	# above glob pattern.
+	@echo "Expect UI assets: $(EXPECT_ASSETS)"
+	@echo "Found UI assets: $(FOUND_ASSETS)"
+
+ifeq ($(FOUND_ASSETS), $(EXPECT_ASSETS))
```

Comment:
I'm open to suggestions of a better approach to checking if `build-ui` was previously run.

Reply:
This is what Make was built for - we can to define the files as dependencies of a target. However, the trick is that the files themselves should also be dependency-tracked. I.e. something like this:
```
placeholder/gen_assets.go: placeholder/public/index.html
    esc ...

actual/gen_assets.go: jaeger-ui/packages/jaeger-ui/build/index.html
    esc ...

build-ui: placeholder/gen_assets.go actual/gen_assets.go
    # actually, nothing to do here
```

But, there are still two problems with this:
  * ideally, `jaeger-ui/packages/jaeger-ui/build/index.html` should also be _dependent_ on the actual UI source. This gets pretty complicated to manage.
  * this set-up _always_ requires a real UI build, a situation I would rather avoid. It should only be required when the binary is compiled with `-tags ui`.

Reply:
Ah yes, good suggestion re: make on file dependencies.

> ideally, jaeger-ui/packages/jaeger-ui/build/index.html should also be dependent on the actual UI source. This gets pretty complicated to manage.

Could we solve this with the following make target spec? I suspect you are suggesting that there's more to consider that I haven't picked up on:
```
jaeger-ui/packages/jaeger-ui/build/index.html:
	cd jaeger-ui && yarn install --frozen-lockfile && cd packages/jaeger-ui && yarn build
```

> this set-up always requires a real UI build, a situation I would rather avoid.

Forgive me for not understanding and my ignorance; what specifically about this set-up requires a real UI build? I thought `yarn build` would result in a real UI build?

> It should only be required when the binary is compiled with -tags ui.

In this PR, I made sure the `build-ui` target is only executed for binaries compiled `-tags ui`, or have I misunderstood something?

Reply:
> Could we solve this with the following make target spec?

The main idea of Make is to execute targets when dependencies change. When the target itself is a file AND has no dependencies, then once that file exists the command will never be run:
```
$ cat Makefile
x: file

file:
       echo n > file

$ make x
echo n > file

$ make x
make: Nothing to be done for `x'.
```

Reply:
We could add the `build` dir to the `clean` target (which doesn't exist when initing the submodule, also note `actual/gen_assets.go` isn't checked in so it should be okay to remove). So we'd have the following:
```
.PHONY: clean
clean:
	rm -rf cover.out .cover/ cover.html lint.log fmt.log \
		cmd/query/app/ui/actual/gen_assets.go \
		jaeger-ui/packages/jaeger-ui/build

.PHONY: build-ui
build-ui: cmd/query/app/ui/actual/gen_assets.go cmd/query/app/ui/placeholder/gen_assets.go
        # Do nothing

jaeger-ui/packages/jaeger-ui/build/index.html:
	cd jaeger-ui && yarn install --frozen-lockfile && cd packages/jaeger-ui && yarn build

cmd/query/app/ui/actual/gen_assets.go: jaeger-ui/packages/jaeger-ui/build/index.html
	esc -pkg assets -o cmd/query/app/ui/actual/gen_assets.go -prefix jaeger-ui/packages/jaeger-ui/build jaeger-ui/packages/jaeger-ui/build

cmd/query/app/ui/placeholder/gen_assets.go: cmd/query/app/ui/placeholder/public/index.html
	esc -pkg assets -o cmd/query/app/ui/placeholder/gen_assets.go -prefix cmd/query/app/ui/placeholder/public cmd/query/app/ui/placeholder/public
```

From a fresh clone, I performed the following test:
```
$ git submodule update --init --recursive

$ make build-ui
cd jaeger-ui && yarn install --frozen-lockfile && cd packages/jaeger-ui && yarn build
esc -pkg assets -o cmd/query/app/ui/actual/gen_assets.go -prefix jaeger-ui/packages/jaeger-ui/build jaeger-ui/packages/jaeger-ui/build
esc -pkg assets -o cmd/query/app/ui/placeholder/gen_assets.go -prefix cmd/query/app/ui/placeholder/public cmd/query/app/ui/placeholder/public

$ make build-ui
# Do nothing

# Trigger a rebuild of UI assets
$ make clean
rm -rf cover.out .cover/ cover.html lint.log fmt.log \
		cmd/query/app/ui/actual/gen_assets.go \
		jaeger-ui/packages/jaeger-ui/build

$ make build-ui
cd jaeger-ui && yarn install --frozen-lockfile && cd packages/jaeger-ui && yarn build
esc -pkg assets -o cmd/query/app/ui/actual/gen_assets.go -prefix jaeger-ui/packages/jaeger-ui/build jaeger-ui/packages/jaeger-ui/build
```

Reply:
Sounds good. I would change `# Do nothing` to `# Do nothing. If you need to force a rebuild of UI assets, run make clean.`

I don't think you need to include `cmd/query/app/ui/actual/gen_assets.go` in the `clean` target, because that file will be rebuilt any time index.html is changed (which is being deleted in clean).